api/handlers/videostream.go

```diff
@@ -35,54 +35,350 @@ LICENSE
+		Results: results,
 		Offset:  qry.Offset,
 		Limit:   qry.Limit,
-		Total:   0,
+		Total:   len(results),
+	})
+}
+
+// CreateVideoStream creates a new video stream.
+// BUG: start and end time are required but this is not being enforced.
+func CreateVideoStream(ctx *fiber.Ctx) error {
+
+	// Parse body.
+	var body CreateVideoStreamBody
+
+	err := ctx.BodyParser(&body)
+	if err != nil {
+		return api.InvalidRequestJSON()
 	}
-	return ctx.JSON(result)
+
+	// Create video stream entity and add to the datastore.
+	cs := model.VideoStream{
+		StartTime:     body.StartTime,
+		EndTime:       &body.EndTime,
+		StreamUrl:     body.StreamUrl,
+		CaptureSource: body.CaptureSource,
+	}
+
+	// Verify CaptureSource exists.
+	store := ds_client.Get()
+	key := store.IDKey("CaptureSource", body.CaptureSource)
+	var captureSource model.CaptureSource
+	if store.Get(context.Background(), key, &captureSource) != nil {
+		return api.DatastoreReadFailure()
+	}
+
+	// Get a unique ID for the new video stream.
+	key = store.IncompleteKey("VideoStream")
+
+	// Add to datastore.
+	key, err = store.Put(context.Background(), key, &cs)
+	if err != nil {
+		print(err.Error())
+		return api.DatastoreWriteFailure()
+	}
+
+	// Return ID of created video stream.
+	id := int(key.ID)
+	return ctx.JSON(VideoStreamResult{
+		ID: &id,
+	})
+}
+
+// StartVideoStream creates a new video stream at the current time.
+func StartVideoStream(ctx *fiber.Ctx) error {
+	// Get time immediately.
+	now := time.Now()
+
+	// Parse body.
+	var body StartVideoStreamBody
+
+	err := ctx.BodyParser(&body)
+	if err != nil {
+		return api.InvalidRequestJSON()
+	}
+
+	// Create video stream entity and add to the datastore.
+	cs := model.VideoStream{
+		StartTime:     now,
+		StreamUrl:     body.StreamUrl,
+		CaptureSource: body.CaptureSource,
+	}
+
+	// Verify CaptureSource exists.
+	store := ds_client.Get()
+	key := store.IDKey("CaptureSource", body.CaptureSource)
+	var captureSource model.CaptureSource
+	if store.Get(context.Background(), key, &captureSource) != nil {
+		return api.DatastoreReadFailure()
+	}
+
+	// Get a unique ID for the new video stream.
+	key = store.IncompleteKey("VideoStream")
+
+	// Add to datastore.
+	key, err = store.Put(context.Background(), key, &cs)
+	if err != nil {
+		print(err.Error())
+		return api.DatastoreWriteFailure()
+	}
+
+	// Return ID of created video stream.
+	id := int(key.ID)
+	return ctx.JSON(VideoStreamResult{
+		ID: &id,
+	})
+}
+
+// EndVideoStream updates the video stream's duration.
+func EndVideoStream(ctx *fiber.Ctx) error {
+	// Get time immediately.
+	now := time.Now()
+
+	// Parse URL.
+	id, err := strconv.ParseInt(ctx.Params("id"), 10, 64)
+	if err != nil {
+		return api.InvalidRequestURL()
+	}
+
+	// Update data in the datastore.
+	store := ds_client.Get()
+	key := store.IDKey("VideoStream", id)
+	var videoStream model.VideoStream
+
+	if store.Update(context.Background(), key, func(e datastore.Entity) {
```

Comment:
It would probably look nicer to do the call to store.Update above.

Reply:
It uses values from the code above, so it must be done in this order

Reply:
Sorry, I just mean above the if statement, so that it's not too crammed in the if condition.